R/messages.R

```diff
@@ -110,3 +110,64 @@ setvar <- function(core, dates, var, values, unit)
 
     invisible(NULL)
 }
+
+
+#' Fetch all the available Hector variables
+#'
+#' This function is similar to fetchvars, except it fetches all available
+#' Hector variables, including exogenous inputs (e.g., emissions & pre-industrial
+#' concentrations), parameters (e.g., alpha, beta, ECS), and output
+#' variables (e.g., concentrations & forcings).
+#'
+#' @param core Hector core object
+#' @param dates Vector of dates, optional
+#' @param scenario Scenario name, optional str
+#' @param outfile Absolute path, including the name, of the output .csv file.
+#' @return vars_all    Dataframe containing all Hector variables
+#' @family main user interface functions
+#' @export
+fetchvars_all <- function(core, dates=NULL, scenario=NULL, outfile=NULL)
+{
+
+    # Get output for variables that do use the date param
+    rslt_date <- fetchvars(core, dates, vars_date, scenario)
+
+    # Get variables that DO NOT use date arg
+    rslt_nodate <- fetchvars(core, NA, vars_nodate, scenario)
+
+    rslt <- rbind(rslt_date, rslt_nodate)
+
+    if (!is.null(outfile)) {
+        # Check that outfile ends with the .csv file extension
+        if (substr(outfile, nchar(outfile)-3, nchar(outfile)) != ".csv") {
+            stop("'outfile' must end in '.csv'")
+        }
+
+        # Pivot the dataframe "wide" so years run along the x axis
+        # Catch warning from reshape()
+        suppressWarnings(rslt_pretty <- reshape(rslt, direction="wide",
```

Comment:
👍 to namespace-qualified `stats::reshape`.

I looked into this and the warning is that there are records with missing time values. Basically, `reshape` doesn't like the fact that all the `rslt_nodate` rows have `NA` for the time variable. The easiest (albeit hacky) way to just circumvent the warning is to give all the `rslt_nodate` rows a fake date value -- e.g. `rslt[is.na(rslt$year), "year"] <- -999`. However, the underlying issue is that there isn't a good wide tabular data structure to efficiently store both time-dependent and time-independent variables without a bunch of `NA`s. In a very wide format like this one, where each row is a variable and each column is a year, you end up with hundreds of columns of `NA`s in rows with time-independent variables. On the other hand, if you spread the variables instead...

```r
reshape(rslt[, colnames(rslt) != "units"], direction = "wide", timevar = "variable", idvar = c("scenario", "year"))
```

...you end up with each parameter column having only one non-`NA` value. If we refactor this as a separate `hector2wide` function (per Robert's suggestion below), it may be worth including a more informative warning about trying to reshape data with missing dates.

Re: `tidyr::pivot_wider` -- I definitely think it's a much better interface, but I'm somewhat reluctant to make it a dependency for just one function, especially given the propensity of the Tidyverse folks to introduce breaking changes (and given that it has 12 dependencies of its own, including some pretty beefy ones). `stats::reshape` is clunky and confusing, with counterintuitive argument names and limited flexibility...but it does get the job done for this specific and relatively-straightforward case.

Reply:
Interesting, thanks.  I think this is a case where the warning is definitely telling us that we're doing something we shouldn't be doing, viz., defining a wide format where many of the variables don't have the coordinate that will be running across the columns.  We should give some thought to what it is that we really want from this operation.  I suspect that what we are producing now will not fully satisfy the users who want wide-format data.